diagnostics: Show how to override default handlers in docs (WAS: Only show one sign per severity per line)

gpanders opened the pull request:
Closes #15770.

Also include documentation on how to customize/replace the built-in decoration/display handlers.

gpanders committed:
test(diagnostic): add test case for signs

gpanders committed:
docs(diagnostic): add blurb on how to replace builtin handlers

justinmk commented on runtime/doc/diagnostic.txt:
```diff
@@ -196,6 +196,28 @@ DiagnosticsChanged	After diagnostics have changed.
 	autocmd User DiagnosticsChanged lua vim.diagnostic.setqflist({open = false })
 <
+==============================================================================
+CUSTOMIZATION
```
*diagnostic-config*

mjlbach commented:
cc'ing @tjdevries since this is changing the defaults, and he commented about the default behavior the last time this was discussed in https://github.com/neovim/neovim/pull/13763

gpanders commented:
I reverted the change limiting signs to one per severity per line, so we'll retain the current default behavior. This PR is simply a documentation change now (I also updated the test case for signs, which I've left in here since it's useful to have).

gpanders committed:
refactor(diagnostics): always make 'set' go through 'show'

Always make calls to `vim.diagnostic.set` call `vim.diagnostic.show`.
This creates an easier to reason about code path and is also less
surprising when users wish to override override `vim.diagnostic.show`
with custom behavior and `vim.diagnostic.set` is called with empty
diagnostics.

Functionally, the end result is the same: when `show` is called with an
empty diagnostics list, it just calls `hide` and then returns, which is
exactly what `reset` does right now.
